### ISSUE TITLE
Metric already exists with Actors

### ISSUE THREAD

**Issue by mrmeyers99** [erikvanoosten/metrics-scala#69]
We use Akka in our project and we create gauges inside our actor.  This causes a problem when the actor dies and restarts.  When it restarts we get an exception that the gauge already exists.  I have the following test that shows the behaviour.

```
import akka.actor.SupervisorStrategy.Restart
import akka.actor._
import akka.testkit.{DefaultTimeout, ImplicitSender, TestKit}
import org.scalatest.{BeforeAndAfterAll, Matchers, WordSpecLike}

import scala.concurrent.duration._
/**
  * a Test to show some TestKit examples
  */
class ActorGuageSpec
  extends TestKit(ActorSystem("ActorGuageSpec"))
  with DefaultTimeout with ImplicitSender with WordSpecLike with Matchers with BeforeAndAfterAll {

  override def afterAll {
    shutdown()
  }

  val supervisor = system.actorOf(Props[SupervisingActor])
  supervisor ! Props[TestActor]
  val actor = expectMsgType[ActorRef]

  "The Actor" should {
    "Be able to replace the guage when it restarts" in {
      within(5.seconds) {
        actor ! 'increase
        expectMsg(1)
        actor ! Kill
        actor ! 'increase
        expectMsg(1)
      }
    }
  }
}

class SupervisingActor extends Actor {
  override val supervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1.second) {
      case _ => Restart
    }

  def receive = {
    case p: Props => sender ! context.actorOf(p)
  }
}


class TestActor extends Actor with Instrumented {

  override val supervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1.second) {
      case _ => Restart
    }

  var counter = 0
  metrics.gauge("test-counter") {
    counter
  }
  override def receive: Receive = {
    case 'increase =>
      counter += 1
      println(counter)
      sender ! counter
  }
}
```

That produces the following log when I run it:

```
Testing started at 5:25 PM ...
1
[ERROR] [12/09/2015 17:25:38.993] [ActorGuageSpec-akka.actor.default-dispatcher-2] [akka://ActorGuageSpec/user/$a/$a] Kill (akka.actor.ActorKilledException)
[ERROR] [12/09/2015 17:25:38.997] [ActorGuageSpec-akka.actor.default-dispatcher-3] [akka://ActorGuageSpec/user/$a/$a] A metric named TestActor.test-counter already exists
akka.actor.PostRestartException: exception post restart (class akka.actor.ActorKilledException)
	at akka.actor.dungeon.FaultHandling$$anonfun$6.apply(FaultHandling.scala:249)
	at akka.actor.dungeon.FaultHandling$$anonfun$6.apply(FaultHandling.scala:247)
	at akka.actor.dungeon.FaultHandling$$anonfun$handleNonFatalOrInterruptedException$1.applyOrElse(FaultHandling.scala:302)
	at akka.actor.dungeon.FaultHandling$$anonfun$handleNonFatalOrInterruptedException$1.applyOrElse(FaultHandling.scala:297)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:36)
	at akka.actor.dungeon.FaultHandling$class.finishRecreate(FaultHandling.scala:247)
	at akka.actor.dungeon.FaultHandling$class.faultRecreate(FaultHandling.scala:76)
	at akka.actor.ActorCell.faultRecreate(ActorCell.scala:374)
	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:464)
	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:483)
	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:282)
	at akka.dispatch.Mailbox.run(Mailbox.scala:223)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.IllegalArgumentException: A metric named TestActor.test-counter already exists
	at com.codahale.metrics.MetricRegistry.register(MetricRegistry.java:91)
	at nl.grons.metrics.scala.MetricBuilder.gauge(MetricBuilder.scala:35)
	at TestActor.<init>(ActorGuageSpec.scala:57)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at java.lang.Class.newInstance(Class.java:442)
	at akka.util.Reflect$.instantiate(Reflect.scala:44)
	at akka.actor.NoArgsReflectConstructor.produce(IndirectActorProducer.scala:105)
	at akka.actor.Props.newActor(Props.scala:214)
	at akka.actor.ActorCell.newActor(ActorCell.scala:562)
	at akka.actor.dungeon.FaultHandling$class.finishRecreate(FaultHandling.scala:234)
	... 11 more
```

I've tried removing guages but that says it's not supported.  What kind of options do we have here?

**Comment by scullxbones** [comment]
Hmm that's interesting.  The library may be able to handle that on your behalf, would need to check into the possibility of only registering the metric if it doesn't already exist.

Usually what I do is keep my metrics in a companion object and reference them from my actors.  This could avoid the problem depending on whether the actor is a singleton or metrics can be shared among actors of the same type - there's bookkeeping here that I can't speak to without knowing more.

**Comment by erikvanoosten** [comment]
The library does first check if your metric already exists, however gauges are different. Gauges are associated with code you wrote. Reporters will continue to call it to get values, even if your actor is dead. What we need is a way to unregister the old gauge. Can we hook into the actor life cycle somehow?

**Comment by mrmeyers99** [comment]
The problem with putting it in an external class is I'm trying to measure the internal state of my object.

**Comment by mrmeyers99** [comment]
Thanks @erikvanoosten.  I was able to get it to work by adding the following to TestActor.

``` 
  @throws[Exception](classOf[Exception])
  override def preRestart(reason: Throwable, message: Option[Any]): Unit = {
    metricRegistry.remove("TestActor.test-counter")
    super.preRestart(reason, message)
  }
```

I had tried metricRegistry.getGauges().remove("TestActor.test-counter"), but that didn't work.

**Comment by mrmeyers99** [comment]
Or a more generic solution:

```
class ActorGaugeSpec
  extends TestKit(ActorSystem("ActorGuageSpec"))
  with DefaultTimeout with ImplicitSender with WordSpecLike with Matchers with BeforeAndAfterAll {

  override def afterAll {
    shutdown()
  }

  val supervisor = system.actorOf(Props[SupervisingActor])
  supervisor ! Props[TestActor]
  val actor = expectMsgType[ActorRef]

  "The Actor" should {
    "Be able to replace the guage when it restarts" in {
      within(5.seconds) {
        actor ! 'increase
        expectMsg(1)
        actor ! Kill
        actor ! 'increase
        expectMsg(1)
      }
    }
  }
}

class SupervisingActor extends Actor {
  override val supervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 10, withinTimeRange = 1.second) {
      case _ => Restart
    }

  def receive = {
    case p: Props => sender ! context.actorOf(p)
  }
}


class TestActor extends InstrumentedActor {

  override val supervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 2, withinTimeRange = 1.second) {
      case _ => Restart
    }

  var counter = 0
  metrics.gauge("test-counter") {
    counter
  }
  override def receive: Actor.Receive = {
    case 'increase =>
      counter += 1
      println(counter)
      sender ! counter
  }
}

trait InstrumentedActor extends Actor with Instrumented {
  @throws[Exception](classOf[Exception])
  override def preRestart(reason: Throwable, message: Option[Any]): Unit = {
    val className = this.getClass.getName
    metricRegistry.getGauges(new MetricFilter {
      override def matches(name: String, metric: Metric): Boolean = name.startsWith(className)
    }).keySet().asScala.foreach(metricRegistry.remove)
    super.preRestart(reason, message)
  }
}
```

**Comment by erikvanoosten** [comment]
Thanks, I was going to suggest something like that ☺

Brian, would you like to take a stab at integrating this further?

Btw, take a look at the stackable trait pattern. It might help here.

**Comment by scullxbones** [comment]
Sure thing @erikvanoosten - I'll take a look.

**Comment by erikvanoosten** [comment]
Off the train now :)
To use the stackable trait pattern you only need to change the definition of `preStart` in `InstrumentedActor` to `abstract override def preStart(...`.

We have to be carefull with filtering by classname though. IMHO it is better to keep an explicit reference to the gauge name(s) and remove them through `metricsRegistry.removeMatching(....)`. Bonus: you can use the implicit conversion in `nl.grons.metrics.scala.Implicits._` :)

Thanks Brian!

**Comment by mrmeyers99** [comment]
Yeah, I was going to suggest something like that.  My solution above would only work if you were to only have one actor of that class, which was true in my case.  You want to create your metrics with the actor path in the name and then you could use the path instead of the class name for your filter.

**Comment by erikvanoosten** [comment]
@scullxbones did you get a chance to look at this problem?

**Comment by erikvanoosten** [comment]
But lets have xmas first ;)

Happy holidays!

**Comment by scullxbones** [comment]
@mrmeyers99 @erikvanoosten please take a look at PR #70 and let me know your thoughts

**Comment by ymeymann** [comment]
FWIW I had the same problem as @mrmyers99 and solved it by removing a gauge if it exists (not very pretty, but works), please see https://github.com/ListnPlay/RiverSong/blob/master/src/main/scala/com/featurefm/riversong/metrics/Instrumented.scala 

It would be great to have a solution in the library, so we could get rid of the wrapper.

**Comment by erikvanoosten** [comment]
Available in 3.5.3 which was just release.
Please let me know how it goes...
